# University Study Prediction Menggunakan SVM

 - Data yang diambil adalah data mahasiswa UPH yang telah lulus 
 - Data yang diperoleh berupa nim, jurusan, nem SMA, SMA asal, sks, dan ipk
 - Prediksi ini menggunakan SVM
 
## Import Package

import package yang diperlukan

In [73]:
# Splitting the dataset into the Training set and Test set 
import random
random.seed(123)
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

## Prepare data

Membaca data dari file datamshclean.csv.

Mengambil kolom kdjur sebagai kdjur dan kdsla,nem,sks,ipk sebagai data2

Menampilkan data yang akan digunakan

Target yang digunakan yaitu kdjur

In [89]:
#read data dr csv
data=pd.read_csv(r"C:\Users\Angel\Downloads\datamhsclean.csv")

#select kolom kdjur pada data jadiin variable kdjur
kdjur= data[['kdjur']]

#data2 ambil beberapa kolom dr data
data2 =data [['kdsla','nem','sks','ipk']]

#ambil dr data2 yang numeric
data2 = data2._get_numeric_data()

#print data yang ada sebanyak 5 baris
data.head(5)

no           nim  kdjur              nmjur     kdsla    nem  sks   ipk  \
0   1  2.090000e+12     71  PENDIDIKAN DOKTER  31851007  40.70  2.0  3.19   
1   2  2.090000e+12     71  PENDIDIKAN DOKTER  27310001  48.95  2.0  3.30   
2   3  2.090000e+12     71  PENDIDIKAN DOKTER  31853015  49.00  2.0  3.23   
3   4  2.090000e+12     71  PENDIDIKAN DOKTER  31875002  49.75  2.0  3.40   
4   6  2.090000e+12     71  PENDIDIKAN DOKTER  31843008  44.35  3.0  3.30   

   status  nemgroup                                     nmsla  
0       0         3             SMA Stella Duce 1, Yogyakarta  
1       0         4                             SMAN 1, Ambon  
2       0         4  SMA Citischool (Satria Wacana), Semarang  
3       0         4                         SMAN 1, Samarinda  
4       0         3         SMA Kristen Bina Bakti 2, Bandung

## Splitting

data akan di pisah menjadi training dan test dimana jumlah test data sebanyak 25% dan training data sebanyak 75%

Menampilkan hasil data yang sudah di train dan test

In [91]:
#split dataset jd training & test
data_train,data_test,kdjur_train,kdjur_test=train_test_split(data2,kdjur,test_size=0.25, random_state=123)

#tampilin yg udah di training & test
print("\ndata_train:\n")
print(data_train.head())
print(data_train.shape)

print("\nkdjur_train:\n")
print(kdjur_train.head())
print(kdjur_train.shape)

print("\ndata_test:\n")
print(data_test.head())
print(data_test.shape)

print("\nkdjur_test:\n")
print(kdjur_test.head())
print(kdjur_test.shape)


data_train:

         nem    sks   ipk
9182   51.70  133.0  2.64
7852   47.40  124.0  3.48
8672   32.00  131.0  3.36
14604  46.25  152.0  3.65
11286  43.60  144.0  2.74
(12039, 3)

kdjur_train:

       kdjur
9182      22
7852      51
8672      43
14604     51
11286     82
(12039, 1)

data_test:

         nem    sks   ipk
6712   54.80  113.0  3.90
2647   54.20   57.0  3.81
6976   50.30  115.0  3.59
9017   49.80  132.0  3.89
10453  43.85  140.0  3.28
(4014, 3)

kdjur_test:

       kdjur
6712      43
2647     113
6976      51
9017      11
10453     61
(4014, 1)


Dapat dilihat 12039 dan 4014 merupakan jumlah data pada train dan test. Angka 1 dan 3 menunjukan banyak kolom pada data

## Feature Scaling

Scaling diperlukan untuk menghilangkan pengaruh feature yg memiliki range besar mendominasi feature yg memiliki range kecil yang biasa disebut dengan normalisasi

In [93]:
# Feature Scaling
from sklearn.preprocessing import StandardScaler

scale_data= StandardScaler()

data_train = scale_data.fit_transform(data_train)
data_test = scale_data.transform(data_test)

#print data train&test yang udah di scaling
print(data_train)
print(data_test)

[[ 0.99318706  0.52888623 -1.32812931]
 [ 0.03550339  0.32271169  0.75707527]
 [-3.39434043  0.48306967  0.45918891]
 ...
 [-0.97785955  0.80378562 -0.06211224]
 [-0.58810457  1.30776783  0.60813209]
 [ 0.61456793 -0.868519   -0.21105542]]
[[ 1.68361016  0.07072058  1.79967757]
 [ 1.54997989 -1.21214323  1.57626279]
 [ 0.68138307  0.11653715  1.03013778]
 ...
 [ 0.58116036  0.94123531  1.35284801]
 [-1.27852768  1.00996016  0.26059799]
 [-0.18721374  0.52888623  0.75707527]]


## Fitting Classifier To The Training Set

Melakukan import package yang diperlukan

Menggunakan kernel linear

Fitting data yang telah di train pada tahap sebelumnya

In [98]:
#import utk fitting classifier to the training set
from sklearn.svm import SVC
from sklearn import svm
from sklearn.model_selection import train_test_split

# Create a linear SVM classifier
clf = SVC(kernel ='linear', random_state=0)

# Train classifier (fitting data yg udh di train)
clf.fit(data_train, kdjur_train.values.ravel())

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
  kernel='linear', max_iter=-1, probability=False, random_state=0,
  shrinking=True, tol=0.001, verbose=False)

## Prediction

In [115]:
#prediction test result
kdjur_pred=clf.predict(data_test)
print (kdjur_pred)

[71 71 71 ... 11 11 11]


## Confusion Matrix

In [116]:
#import confusion matrix
from sklearn.metrics import confusion_matrix

#confusion matrix
cm = confusion_matrix(kdjur_test, kdjur_pred)
print(cm)

[[627   0   0   0   0   0   0   0   0   0   0   0   0   0  98   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0]
 [141   0   0   0   0   0   0   0   0   0   0   0   0   0  78   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0]
 [ 72   0   0   0   0   0   0   0   0   0   0   0   0   0   8   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0]
 [146   0   0   0   0   0   0   0   0   0   0   0   0   0  26   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0]
 [141   0   0   0   0   0   0   0   0   0   0   0   0   0  30   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0]
 [114   0   0   0   0   0   0   0   0   0   0   0   0   0  20   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0]
 [ 26   0   0   0   0   0   0   0   0   0   0   0   0   0   7   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0]
 [ 10   0   0   0   0   0   0   0   0   0   0   0   0   0   3   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0]


## K Fold

accuracy digunakan untuk mengevaluasi model yang merupakan rasio dari jumlah kejadian yang diprediksi dengan benar dibagi dengan jumlah total kasus dalam dataset untuk memberikan persentase.  Kita akan menggunakan variabel penilaian saat menjalankan build dan mengevaluasi setiap model di langkah selanjutnya

In [126]:
#K Fold
from sklearn import model_selection
from sklearn.model_selection import cross_val_score

kfold = model_selection.KFold(n_splits=10, random_state=7)

modelCV = SVC(kernel ='linear', random_state=0)
scoring ='accuracy'
results = model_selection.cross_val_score(modelCV,data_train,kdjur_train.values.ravel(), cv=kfold, scoring=scoring)
print("cross fold validation average accuracy: %.3f"%(results.mean()))



cross fold validation average accuracy: 0.252


## Evaluating Classification Report

In [118]:
#evaluating classification report
from sklearn.metrics import classification_report
print(classification_report(kdjur_test, kdjur_pred))

              precision    recall  f1-score   support

          11       0.21      0.86      0.34       725
          12       0.00      0.00      0.00       219
          21       0.00      0.00      0.00        80
          22       0.00      0.00      0.00       172
          23       0.00      0.00      0.00       171
          24       0.00      0.00      0.00       134
          25       0.00      0.00      0.00        33
          32       0.00      0.00      0.00        13
          33       0.00      0.00      0.00        99
          34       0.00      0.00      0.00       212
          41       0.00      0.00      0.00       196
          43       0.00      0.00      0.00       164
          51       0.00      0.00      0.00       331
          61       0.00      0.00      0.00       210
          71       0.38      0.72      0.50       552
          81       0.00      0.00      0.00        48
          82       0.00      0.00      0.00        72
          83       0.00    

C:\Users\Angel\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [125]:
import pickle

# write python dict to a file
#mydict = {'a': 1, 'b': 2, 'c': 3}
output = open('kdjur_pred.pkl', 'wb')
pickle.dump(kdjur_pred, output)
output.close()

# read python dict back from the file
#pkl_file = open('data_train.pkl', 'rb')
#mydict2 = pickle.load(pkl_file)
#pkl_file.close()

#print (data_train)
#print (mydict2)